README.md
# Data Incremental
Data Incremental is a game about storing data and mastering more advanced technologies as you ascend the order of magnitudes!

This game works from LÖVE, so you'll have to install LÖVE from [this link](https://love2d.org/), download Data Incremental's source code (Preferrably ZIP), and drag the folder "Data Incremental" onto love.exe file (which should be in C:/Program Files/LOVE).

Data Incremental supports only 600x800 resolution right now, 1600x900 coming soon!


changelog.txt
v0.1 (20.10.2023)
    Added 3 first Bytes upgrades: Binary Encoder, Multi-processing and Data Cascade. Each next upgrade boosts the previous one.

    Added Efficiency, which boosts Bytes gain at the cost of your system. When you store more than 1e6 Bytes, you can reset all Bytes upgrades
    and Bytes in exchange for Efficiency. Efficiency is measured in percentage, each % corresponds to +×0.01 boost to Bytes. The initial
    Efficiency is 100% (×1).

v0.1.1 (22.10.2023)
    Fixed some bugs.

main.lua
--[[ Names of upgrades:
Binary Encoder - costs 8 bytes
Multi-processing - costs 512 bytes
Data Cascade - costs 8192 bytes
--]]

require "mousepressed"
require "BinaryEncoder"
require "MultiProcessing"
require "DataCascade"
require "BottomPanel"
require "Efficiency"

love.window.setTitle("Data Incremental v0.1")

points = 8
screen = "bytes"

font = love.graphics.setNewFont(16)

function love.draw()
    love.graphics.setColor(1, 1, 1)
    font = love.graphics.setNewFont(16)
    if points < 1000000 then
    love.graphics.printf(string.format("A total of %.2f bytes are stored.", points), 150, 10, 500, "center")
    else
        love.graphics.printf(string.format("A total of %.2e bytes are stored.", points):gsub("e[%+0%-0]+", "e"), 150, 10, 500, "center")
    end
    love.mousepressed(x, y, button, istouch, presses)
    BinaryEncoderDisplay()
    MultiProcessingDisplay()
    DataCascadeDisplay()
    BottomPanelDisplay()
    EfficiencyDisplay()
end

function love.update(dt)
    points = points + BinaryEncoderEffect * dt
    BinaryEncoderDisplay()
    MultiProcessingDisplay()
    DataCascadeDisplay()
    BottomPanelDisplay()
    EfficiencyDisplay()
end

Efficiency.lua
Efficiency = 100
EfficiencyOnReset = 0

function EfficiencyDisplay()
    EfficiencyOnReset = (math.log(points + 1) ^ 2.4) / 54.56
    if screen == "efficiency" then
        love.graphics.setColor(1, 1, 1)
        love.graphics.rectangle("line", 170, 160, 120, 40)
        love.graphics.setColor(0.2, 0.2, 0.2)
        love.graphics.rectangle("fill", 171, 161, 118, 38)
        love.graphics.setColor(1, 1, 1)
        love.graphics.print("Convert", 202, 172)
        if points > 1000000 then
            love.graphics.printf(string.format("+%.2f%% Efficiency upon conversion.", EfficiencyOnReset), 300, 172, 400, "left")
        else
            love.graphics.printf(string.format("Get %.0f more Bytes to be able to convert.", 1000000 - points), 300, 172, 400, "left")
        end
        love.graphics.printf({{1, 1, 1}, "Your system's Efficiency is ", {0, 0.68, 1}, string.format("%.2f%%", Efficiency), {1, 1, 1}, string.format(", boosting Bytes gain by x%.2f.", Efficiency/100)}, 100, 60, 600, "center")
        love.graphics.printf(string.format("After storing more than 1e6 Bytes, you can reset your system and convert your Bytes for additional Efficiency, further boosting Bytes gain."), 200, 85, 400, "center")
    end
end

DataCascade.lua
DataCascadePrice = 8192
DataCascadePower = 1
DataCascadeEffect = 1
DataCascadePurchases = 0

function DataCascadeDisplay()
    if screen == "bytes" then
        love.graphics.setLineWidth(2)
        love.graphics.setLineStyle("rough")

        love.graphics.setColor(1, 1, 1)
        love.graphics.rectangle("line", 200, 220, 150, 30)
        love.graphics.setColor(0.2, 0.2, 0.2)
        love.graphics.rectangle("fill", 201, 221, 148, 28)
        love.graphics.setColor(1, 1, 1)

        font = love.graphics.setNewFont(14)
        love.graphics.printf("Data Cascade (" .. DataCascadePurchases .. ")", 0, 227, 185, "right")
        font = love.graphics.setNewFont(12)

        if DataCascadePrice < 1000000 then -- 1e6 is threshold for scientific notation --
            love.graphics.printf(string.format("Cost: %.2f bytes", DataCascadePrice), 200, 227, 150, "center")
        else
            love.graphics.printf(string.format("Cost: %.2e bytes", DataCascadePrice):gsub("e[%+0%-0]+", "e"), 200, 227, 150, "center")
        end

        love.graphics.printf("Power: +×" .. DataCascadePower .. " to Multi-Processing", 160, 255, 1600, "left")
        love.graphics.printf("Effect: ×" .. DataCascadeEffect .. " to Multi-Processing", 160, 270, 1600, "left")
    end
end

mousepressed.lua
function love.mousepressed(x, y, button, istouch, presses)
    if button == 1 then
        if screen == "bytes" then
            if (x > 200 and x < 350) and (y > 50 and y < 80) then
                if points >= BinaryEncoderPrice then -- Binary Encoder buying --
                    points = points - BinaryEncoderPrice
                    BinaryEncoderPrice = BinaryEncoderPrice * 1.5
                    BinaryEncoderPurchases = BinaryEncoderPurchases + 1
                    BinaryEncoderEffect = BinaryEncoderPurchases * BinaryEncoderPower
                end

            elseif (x > 200 and x < 350) and (y > 135 and y < 165) then
                if points >= MultiProcessingPrice then -- Multi-processing buying --
                    points = points - MultiProcessingPrice
                    MultiProcessingPrice = MultiProcessingPrice * 1.5
                    MultiProcessingPurchases = MultiProcessingPurchases + 1
                    MultiProcessingEffect = 1 + MultiProcessingPurchases * MultiProcessingPower
                    BinaryEncoderPower = MultiProcessingEffect
                    BinaryEncoderEffect = BinaryEncoderPurchases * BinaryEncoderPower
                end

            elseif (x > 200 and x < 350) and (y > 220 and y < 250) then
                if points >= DataCascadePrice then -- Data Cascade buying --
                    points = points - DataCascadePrice
                    DataCascadePrice = DataCascadePrice * 1.5
                    DataCascadePurchases = DataCascadePurchases + 1
                    DataCascadeEffect = 1 + DataCascadePurchases * DataCascadePower
                    MultiProcessingPower = DataCascadeEffect
                    MultiProcessingEffect = 1 + MultiProcessingPurchases * MultiProcessingPower
                    BinaryEncoderPower = MultiProcessingEffect
                    BinaryEncoderEffect = BinaryEncoderPurchases * BinaryEncoderPower
                end
            
            elseif (x > 80 and x < 160) and (y > 569 and y < 599) then
                screen = "efficiency"
            end
        

        elseif screen == "efficiency" then
            if (x > 170 and x < 290) and (y > 160 and y < 200) then
                if points >= 1000000 then
                    Efficiency = Efficiency + EfficiencyOnReset
                    EfficiencyOnReset = 0
                    points = 8

                    DataCascadePurchases = 0
                    DataCascadePower = 1
                    DataCascadeEffect = DataCascadePurchases * DataCascadePower + 1
                    DataCascadePrice = 8192

                    MultiProcessingPurchases = 0
                    MultiProcessingPower = 1
                    MultiProcessingEffect = MultiProcessingPurchases * MultiProcessingPower + 1
                    MultiProcessingPrice = 512

                    BinaryEncoderPurchases = 0
                    BinaryEncoderPower = (Efficiency / 100)
                    BinaryEncoderEffect = BinaryEncoderPurchases * BinaryEncoderPower
                    BinaryEncoderPrice = 8
                end

            elseif (x > 1 and x < 81) and (y > 569 and y < 599) then
                screen = "bytes"
            end
        end
    end
end

BinaryEncoder.lua
BinaryEncoderPrice = 8
BinaryEncoderPower = 1
BinaryEncoderEffect = 0
BinaryEncoderPurchases = 0

function BinaryEncoderDisplay()
    if screen == "bytes" then
        love.graphics.setLineWidth(2)
        love.graphics.setLineStyle("rough")

        love.graphics.setColor(1, 1, 1)
        love.graphics.rectangle("line", 200, 50, 150, 30)
        love.graphics.setColor(0.2, 0.2, 0.2)
        love.graphics.rectangle("fill", 201, 51, 148, 28)
        love.graphics.setColor(1, 1, 1)

        font = love.graphics.setNewFont(14)
        love.graphics.printf("Binary Encoder (" .. BinaryEncoderPurchases .. ")", 0, 57, 185, "right")
        font = love.graphics.setNewFont(12)

        if BinaryEncoderPrice < 1000000 then -- 1e6 is threshold for scientific notation --
            love.graphics.printf(string.format("Cost: %.2f bytes", BinaryEncoderPrice), 200, 57, 150, "center")
        else
            love.graphics.printf(string.format("Cost: %.2e bytes", BinaryEncoderPrice):gsub("e[%+0%-0]+", "e"), 200, 57, 150, "center")
        end

        love.graphics.printf(string.format("Power: +%.2f to Bytes gain", BinaryEncoderPower), 160, 85, 1600, "left")
        love.graphics.printf(string.format("Effect: +%.2f to Bytes gain", BinaryEncoderEffect), 160, 100, 1600, "left")
    end
end

MultiProcessing.lua
MultiProcessingPrice = 512
MultiProcessingPower = 1
MultiProcessingEffect = 1
MultiProcessingPurchases = 0

function MultiProcessingDisplay()
    if screen == "bytes" then
        love.graphics.setLineWidth(2)
        love.graphics.setLineStyle("rough")

        love.graphics.setColor(1, 1, 1)
        love.graphics.rectangle("line", 200, 135, 150, 30)
        love.graphics.setColor(0.2, 0.2, 0.2)
        love.graphics.rectangle("fill", 201, 136, 148, 28)
        love.graphics.setColor(1, 1, 1)

        font = love.graphics.setNewFont(14)
        love.graphics.printf("Multi-Processing (" .. MultiProcessingPurchases .. ")", 0, 142, 185, "right")
        font = love.graphics.setNewFont(12)

        if MultiProcessingPrice < 1000000 then -- 1e6 is threshold for scientific notation --
            love.graphics.printf(string.format("Cost: %.2f bytes", MultiProcessingPrice), 200, 142, 150, "center")
        else
            love.graphics.printf(string.format("Cost: %.2e bytes", MultiProcessingPrice):gsub("e[%+0%-0]+", "e"), 200, 142, 150, "center")
        end

        love.graphics.printf("Power: +×" .. MultiProcessingPower .. " to Binary Encoder", 160, 170, 1600, "left")
        love.graphics.printf("Effect: ×" .. MultiProcessingEffect .. " to Binary Encoder", 160, 185, 1600, "left")
    end
end

BottomPanel.lua
function BottomPanelDisplay()
    font = love.graphics.setNewFont(14)

    -- Highlight Bytes --
    
    if screen == "bytes" then
        love.graphics.setColor(1, 1, 1)
        love.graphics.rectangle("line", 1, 569, 80, 30)
        love.graphics.setColor(0.3, 0.3, 0.3)
        love.graphics.rectangle("fill", 2, 570, 78, 28)
        love.graphics.setColor(1, 1, 1)
    else
        love.graphics.setColor(1, 1, 1)
        love.graphics.rectangle("line", 1, 569, 80, 30)
        love.graphics.setColor(0.2, 0.2, 0.2)
        love.graphics.rectangle("fill", 2, 570, 78, 28)
        love.graphics.setColor(1, 1, 1)
    end
    love.graphics.print("Bytes", 21, 575)

    -- Highlight Efficiency --

    if screen == "efficiency" then
        love.graphics.setColor(1, 1, 1)
        love.graphics.rectangle("line", 80, 569, 80, 30)
        love.graphics.setColor(0.3, 0.3, 0.3)
        love.graphics.rectangle("fill", 81, 570, 78, 28)
        love.graphics.setColor(1, 1, 1)
    else
        love.graphics.setColor(1, 1, 1)
        love.graphics.rectangle("line", 80, 569, 80, 30)
        love.graphics.setColor(0.2, 0.2, 0.2)
        love.graphics.rectangle("fill", 81, 570, 78, 28)
        love.graphics.setColor(1, 1, 1)
    end
    love.graphics.print("Efficiency", 85, 575)

    -- Highlight Layering --

    love.graphics.setColor(1, 1, 1)
    love.graphics.rectangle("line", 160, 569, 80, 30)
    love.graphics.setColor(0.2, 0.2, 0.2)
    love.graphics.rectangle("fill", 161, 570, 78, 28)
    love.graphics.setColor(1, 1, 1)

    love.graphics.print("Layering", 169, 575)
end